### DISCUSSION TITLE
DateTime Literal (Needed Mostly for Tuples)

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#1205] [General]
Why there isn't a DateTime literal in C# after all those years?
C# is introducing many new compact syntax forms but not paying attention to this important, repeatedly needed feature. 
It's more important today because of Tuples. Isn't it nice to write:
`var Period = (Start: #1/1/2017#, End: #1/1/2018#);`

There are two more resons to add the ## literal:
 1- DateTime.Parse adds the current today if the date part is missing. ## in VB.NET adds the min. date 1/1/1.
 2- DateTime.Parse(string s) uses the CultureInfo.CurrentCulture, while ## in VB.NET uses the Invariant Culture ("en"). So, using DateTime.Parse(s), will have different results on different PCs!.. On Windows with Arabic culture, that code will cause an exception, because this culture uses the format: "dd/MM/yyyy" and there is no 15th month! This is a bad error-prone design! so
#1/2/2018# is a shortcut of DateTime.Parse("1/2/2018", new CultureInfo("en")).

**Comment by HaloFour** [comment]
This isn't needed for tuples anymore than literal syntax for any other arbitrary data type:

```cs
var Period = (Start: new DateTime(2017, 1, 1), End: new DateTime(2018, 1, 1));
```

Given that there are already several flavors of date/time data types in the BCL, and [possibly more coming](https://github.com/dotnet/corefxlab/tree/master/src/System.Time), I don't really think this is a good idea.  VB.NET does have Date literals and it's permanently locked to the `System.DateTime` struct, despite it often not being the right tool for the job.

**Comment by mburbea** [comment]
I don't think having a datetime literal would preclude specifying a type. We can always add suffixes like the numeric literals in C#
```c#
var tuple = (@int:3,@uint:3U,@long:3L,@ulong:3UL,@float:3F,@double:3D)
```

or prefixes like ODBC type literals in languages like pg/sql.
```sql
select {ts'2017-12-20'} -- timestamp
          ,{d'2017-12-20'} -- date
```

I personally would enjoy having the option for user-defined literals, but as a minimum I find dates are things I'd like to make constant and be able to use in places for default parameters. VB6 does allow this behavior as it supports datetime literals, but I don't have that parity to do in C#.

**Comment by ghost** [comment]
> several flavors of date/time data types in the BCL, and possibly more coming,

Each of them can have the same literal with different format.
If the goal is only to do things, things already has been done for years, so why you bother introducing new language features such as Tuples?
Date and Time value are used repeatedly, and it's a time waste to send date and time parts to a constructor over and over. 
Besides, the absence of DateTime literals prevents us from declaring a Datetime constant (unless using a readonly field), and makes it hard to pass date/time values to attributes. This is a prime data type but you can't write it!

**Comment by HaloFour** [comment]
@mburbea 

> but as a minimum I find dates are things I'd like to make constant and be able to use in places for default parameters.

That would be a separate feature.  VB.NET doesn't have support for constants or default parameters because it has a literal syntax, but rather because that compiler has extra support for faking those constructs through language-specific attributes.  Note that a VB.NET `Date` constant or default value is not a proper CLR constant or default value, and VB.NET can't work around the fact that the CLR can't support `DateTime` in attributes.

```vb
Public Class C
    Const d As Date = #2017/01/01#

    Public Sub M1(Optional d As Date = #2018/01/01#)
    End Sub
    
    Public Sub M2()
        M1()
    End Sub
End Class

' Actually:

Public Class C
    <DateTimeConstant(636188256000000000L)>
    Private Shared d As DateTime = New DateTime(636188256000000000L)

    Public Sub M1(<DateTimeConstant(636503616000000000L)> Optional d As DateTime=Nothing)
    End Sub

    Public Sub M2()
        Me.M1(New DateTime(636503616000000000L))
    End Sub
End Class
```

But my bigger concern is with baking in the various date/time types into the language.  We have `DateTime`, `DateTimeOffset`, `SqlDateTime`, `Date` and `Time`, not to mention NodaTime, which is what developers probably should be using over anything in the BCL.

@MohammadHamdyGhanem 

> Besides, the absence of DateTime literals prevents us from declaring a Datetime constant (unless using a readonly field), and makes it hard to pass date/time values to attributes.

There's nothing that the C# compiler can do about this.  Attributes support a very small subset of specific data types as arguments because those values have to be embedded as blobs directly into the assembly and the CLR loads them directly.  The language can't influence how that works.  This is why VB.NET, despite having date/time literals and supporting pseudo date/time constants/defaults still can't support date/time attribute properties.

**Comment by HaloFour** [comment]
And when I mention that attributes can't support `DateTime`, etc.  I don't mean that it couldn't be done, just that it doesn't today, and that would require changes to the CLR.

**Comment by ghost** [comment]
> And when I mention that attributes can't support DateTime, etc. I don't mean that it couldn't be done, just that it doesn't today, and that would require changes to the CLR.

So, C# should be ready and add date/time Literals :)
May be C# Team needs voting on this subject on C# blogger or something.

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

> So, C# should be ready and add date/time Literals :)

There's no *technical* reason why C# can't have date/time literals, at least to the extent that VB.NET supports them.  But that's never reason enough to implement a feature.  It's not like the C# team has been ignorant of the fact that VB.NET has date/time literals or how VB.NET went about implementing them.  There might be a reason why C# doesn't have them.

> May be C# Team needs voting on this subject on C# blogger or something.

[UserVoice](https://visualstudio.uservoice.com/forums/121579-visual-studio-ide/category/30931-languages-c) is the venue for voting on proposals like that.  But this isn't a democracy.  In the end it's up to someone from the C# team to "champion" any proposal and for it to be successfully argued in design meetings before it would be considered for adoption.

**Comment by ghost** [comment]
My point of view is we are in th "Pattern Programming" era.  Projects are growing bigger and more complex every day. Language syntax should be smarter, time saver and more readable (but still shorter)!
This is why I posted another propasal about "matches" keyword, despite it saves about 3 lines of code (including the using statement)!
5 seconds less to do a frequent task, can save hours monthly!

**Comment by stakx** [comment]
@MohammadHamdyGhanem:

> Isn't it nice to write:
>
> ```csharp
> var Period = (Start: #1/1/2017#, End: #1/1/2018#);
> ```

If you asked me, the answer would be a resounding "no, this isn't nice". I am not American, and I prefer my dates formatted differently. At the very least, your proposal should use ISO formatted dates (but even that would be debatable).

Dates and times are a complicated topic, thus the existence of whole specialized libraries such as Noda Time. There's almost certainly no way you're going to find one syntax to cater for all different needs when it comes to this topic. Better leave this to dedicated libraries, and reserve literals for unambiguous, simple data types... not complex, culture-specific ones.

> This is whu I posted another propasal about "matches" keyword, despite it saves about 3 lines of code (including the using statement)! 5 seconds less to do a frequent task, can save hours monthly!

Incidentally, your other proposal is a fine example of how spending a little more time on researching the BCL could've saved you the time to write the proposal. 😉

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

I can't find a good existing reference to it, but to paraphrase what members of the LDM have stated on several occasions, if it can be managed easily by a library, either as an existing method or one that you have to write once, then it doesn't belong in the language.  Your regex `matches` keyword is a very good example of something that doesn't belong in the language.  At least not as simply a helper syntax to using the `Regex` class.  Now if you want to argue about compile-time validation of regex patterns that would be something else, and there's already a proposal for it:  #371

**Comment by ghost** [comment]
@stakx 
`I am not American, and I prefer my dates formatted differently`
Me the same, but we are writing code in English, and the format you use to write date/time  in code doesn't have to be the format you display to the end user.

**Comment by ghost** [comment]
@HaloFour 
You can do LinQ with methods, but they gave it another syntax. As I said before, almost every thing can be done from C# 1.0 (except generics)! 
We need to do it, and do it quickly!
The "inline reg. exp. Matching" idea came from our discussion about Constrained Types. If the reg. exp. can appear in type declaration, so it can appear in if statement as well. Make it a prime part of the language, and as other proposed give it some syntax checking.
So, if reg. exp. can appear as literals, why dates can't? All ideas are connected. 
Notice that Microsoft pushes millions of VB.NET programmers to use C# in Xamarin and other new technologies. VB.NET programmers tasted the beauty of VB smart syntax, and are used to choose from different ways to do the same thing. So, if Microsoft is busy to develop VB.NET in those areas, then at least it should make C# syntax as smart as possible, to hear less complaints. C# community will suffer from Microsoft bad VB.NET strategy!

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

> You can do LinQ with methods, but they gave it another syntax.

Yes, which gets unwieldy sometimes with various constructs like group joins.

> As I said before, almost every thing can be done from C# 1.0 (except generics)!

Yes, but when there are specific enhancements

> We need to do it, and do it quickly!

Nothing happens quickly here.  The design process is very conservative.  Every feature must justified the incredibly high cost in development time and support for the life of the language.  The vast majority of little proposals here will never be considered, let alone championed.  Your regex proposal will most likely sit in that same bucket.

Reimplementing one-liners in the BCL as language grammar is not "smart syntax".  It's a massive waste of time and resources better spent on solving problems that would actually benefit developers.

**Comment by HaloFour** [comment]
Ouch, that's curmudgeonly even for me.  Let me backpedal that a bit.

The standard line is that each feature needs 100 points to be considered.  Each starts with -100 points, just because implementing any feature costs time and effort.  The feature has to be good enough to make up the difference, and then some.  Even little features have to demonstrate measurable reward.

A language isn't perfect when there's nothing left to add.  It's perfect when there's nothing left to remove.

**Comment by ufcpp** [comment]
> var Period = (Start: #1/1/2017#, End: #1/1/2018#);

No! DateTime literal should be like 2017/1/1! at least in Japan.

This claim is a bit wild but what do you think about culture problem? Do you want to have culture dependent literals? There are so many confusion between dd/MM/yyyy and MM/dd/yyyy.

**Comment by jnm2** [comment]
I refuse to write dates as anything other than yyyy-MM-dd. That includes handwriting.

**Comment by DavidArno** [comment]
> Me the same, but we are writing code in English, and the format you use to write date/time in code doesn't have to be the format you display to the end user.

I'm English, so as long as `#1/1/2017#` were in `d/m/y` format, I'd be happy 😀. But a great many American developers would swear at it every time they got the month and day the wrong way around.

I actually quite like this suggestion (accepting the caveats around the many date types that exist), but if implemented, I believe it would have to use the ISO format, `#2017-1-1#`.

@jnm2,
I tend to use the month name, eg `1st Jan 2017`, but that presupposes my audience speaks English. Always using `yyyy-MM-dd` seems extreme, but I admire your discipline.

**Comment by ufcpp** [comment]
Actually VB DateTime literal is hard to use for Japanese. Even if one can read English, he/she feels hard to read en-us culture date. Language and culture are different things.

**Comment by DavidArno** [comment]
@ufcpp,

That surprises me as I thought the Japanese date format is the same as the ISO one, `yyyy-MM-dd`, and I though that was what VB.Net used. 😕

**Comment by yaakov-h** [comment]
What if I want my `DateTime` literal to be interpreted in `System.Globalization.HebrewCalendar`, or `JulianCalendar`, or another one of the 10 calendars supported by the BCL that are not the Gregorian one?

What happens if my `DateTime` literal is an invalid date, such as the 2019-02-29? Does the compiler have to keep track of valid calendar dates?

**Comment by DavidArno** [comment]
> What if I want my `DateTime` literal to be interpreted in `System.Globalization.HebrewCalendar`, or `JulianCalendar`, or another one of the 10 calendars supported by the BCL that are not the Gregorian one?

Oh, why does the real world have to be so complicated? 😀 

But that is a very good point. I think my previous lukewarm support for this is rapidly cooling.

**Comment by yaakov-h** [comment]
https://github.com/kdeldycke/awesome-falsehood#dates-and-time

**Comment by ufcpp** [comment]
@DavidArno 

https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/data-types/date-data-type

VB's DateTime literal is US format, M/d/yyyy.

**Comment by DavidArno** [comment]
@ufcpp, well that's just rubbish! 😞 

Thanks for the link.

**Comment by HaloFour** [comment]
@ufcpp @DavidArno 

They've recently added "yyyy-MM-dd" as well, but yeah, for a long time it was hard-coded to the US format.

If C# were ever to adopt date/time literals I would seriously hope that they wouldn't consider any format other than ISO-8601.  Through variations of that format they could support a number of different potential data types as well:

```cs
DateTime dt = #2017-12-21T12:00:00.000#;
DateTime dtUtc = #2017-12-21T12:00:00.000Z#; // target typing
DateTimeOffset dtz = #2017-12-21T12:00:00.000+04:00#;
DateTimeOffset dtzUtc = #2017-12-21T12:00:00.000Z#; // again, target typing
Date d1 = #2017-12-21#;
Date d2 = #2017-353#;
Date d3 = #2017-W51-4#;
Time t = #12:00:00#;
TimeSpan ts = #PT1M30S#;
```

**Comment by svick** [comment]
@yaakov-h 

>What happens if my `DateTime` literal is an invalid date, such as the 2019-02-29? Does the compiler have to keep track of valid calendar dates?

[That's exactly what the VB compiler does.](https://github.com/dotnet/roslyn/blob/57fd934/src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb#L2380-L2392)

**Comment by CyrusNajmabadi** [comment]
It's pretty simple.  In order to be part of the language the feature has to have high demonstrable value.  That means, if you can already do it with a library, the ability to do in in the language proper needs to be a *significantly* better experience than what you can already do today.  If you can't meet that bar, then it's not worth the cost it's going to take to spec and implement all of this.

DateTimes are in that category of things i do *not* want my language getting involved with.  Having seen how that space has played out in numerous frameworks, i have no confidence that anything we did in the language would be good enough to be a viable tool for the user 5, 10, 15 years down the line.

**Comment by aka-nse** [comment]
In proposals which are on discussions, I think literals that really valuable to be embedded in language are only UTF-8 string literal at most. If other literals including `DateTime` are inevitable by all means, a feature like as user defined literal seems just proper. But I don't feel such a great necessity.

**Comment by ghost** [comment]
@HaloFour @mburbea @jnm2 @DavidArno @yaakov-h 
There are two more resons to add the ## literal:
1- DateTime.Parse adds the current today if the date part is missing. ## in VB.NET adds the min. date 1/1/1.
2- DateTime.Parse(string s) uses the CultureInfo.CurrentCulture, while ## in VB.NET uses the Invariant Culture ("en"). So, using DateTime.Parse(s), will have different results on different PCs!.. On Windows with Arabic culture, that code will cause an exception, because this culture uses the format: "dd/MM/yyyy" and there is no 15th month! This is a bad error-prone design! so
#1/2/2018# is a shortcut of DateTime.Parse("1/2/2018", new CultureInfo("en")).

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

1.  `DateTime` is the wrong type to use if you don't actually have a date part.  That's a very good reason to *not* support a literal.
2.  This is why you use ISO-8601 formatted date times and not locale sensitive formats.  VB.NET's support for Invariant culture comes very specifically from legacy code meant to interpret American-style dates, a format largely unknown outside of the US.

**Comment by ghost** [comment]
1- I'm not aware ot time types. Where are they?
2- I use Invariant Culture format. This is done by #s# in VB.NET, and the longer DateTime.Parse(s, new CultureInfo("en")). Is ther a problem in making things easy and simple in C#?

**Comment by jnm2** [comment]
@MohammadHamdyGhanem last item in https://github.com/dotnet/corefxlab#components.

**Comment by theunrepentantgeek** [comment]
You might also find it useful to check out [NodaTime](https://nodatime.org/).

**Comment by TahirAhmadov** [comment]
I would say that while having `DateTime` literals would be nice, `TimeSpan` literals would be even more helpful.

**Reply by HaloFour** [reply]
The existing syntax might be a little verbose but is very clear and easy to use:

```cs
var timespan = TimeSpan.FromSeconds(1234);
```

It's also pretty easy to add extension methods to make them feel like post-fix operations:

```cs
var timespan = 1234.seconds();
```

With extension everything you could make it a property which would be slightly more ergonomic:

```cs
var timespan = 1234.seconds;
```

**Reply by TahirAhmadov** [reply]
The problem with these types of extensions is while they are usable, they add a ton of noise. With types like `object`, `int`, etc., this reduces usability with suggestions and stuff like that.
